src/net_processing.cpp

```diff
@@ -2309,6 +2379,100 @@ bool PeerManagerImpl::CheckHeadersAreContinuous(const std::vector<CBlockHeader>&
+        }
+
+        previously_downloaded_headers.swap(result.headers_to_process);
+
+        if (previously_downloaded_headers.empty() && result.success) {
+            // If nothing else was returned and processing was successful, then
+            // we're all done.
+            return true;
+        }
+    }
+    // Either we didn't have a sync in progress, or something went wrong
+    // processing these headers, or we are returning headers to the caller to
+    // process.
+    return false;
+}
+
+bool PeerManagerImpl::TryLowWorkHeadersSync(Peer& peer, CNode& pfrom, const CBlockIndex* chain_start_header, const std::vector<CBlockHeader>& headers)
+{
+    // Calculate the total work on this chain.
+    arith_uint256 total_work = chain_start_header->nChainWork + CalculateHeadersWork(headers);
+
+    // Our dynamic anti-DoS threshold (minimum work required on a headers chain
+    // before we'll store it)
+    arith_uint256 minimum_chain_work = GetAntiDoSWorkThreshold();
+
+    // Avoid DoS via low-difficulty-headers by only processing if the headers
+    // are part of a chain with sufficient work.
+    if (total_work < minimum_chain_work) {
+        // Only try to sync with this peer if their headers message was full;
+        // otherwise they don't have more headers after this so no point in
+        // trying to sync their too-little-work chain.
+        if (headers.size() == MAX_HEADERS_RESULTS) {
+            // Note: we could advance to the last header in this set that is
+            // known to us, rather than starting at the first header (which we
+            // may already have). But we actually might have all the headers in
+            // this set already, because headers sync can be imprecise when a
+            // peer has to serve us a long chain (due to imprecision in the way
+            // locators are calculated). So philosophically, if we wanted to
+            // optimize this correctly, we could consider requesting more
+            // headers until we find the peer's first new header on this chain,
+            // and start the sync from there. In practice this is unlikely to
+            // matter much outside of initial sync, which we generally only do
+            // once, so for now we'll just start the sync with the first header
+            // in the set and not worry about this issue.
+            peer.m_headers_sync.reset(new HeadersSyncState(peer.m_id, m_chainparams.GetConsensus()));
+            if (std::optional<CBlockLocator> locator =
+                    peer.m_headers_sync->StartInitialDownload(chain_start_header,
+                        headers, minimum_chain_work,
+                        m_chainman.ActiveChain().GetLocator(chain_start_header)))
+            {
+                Assume(!locator->vHave.empty());
+                if (!locator->vHave.empty() && MaybeSendGetHeaders(pfrom, *locator, peer)) {
+                    LogPrint(BCLog::NET, "more getheaders (from %s) to end to peer=%d\n",
```

Comment:
5691c174c5cb2f9947e6cc2db8d32b4b4715e47d : can you add the height to this message? That and a message when the state changes, makes it easier to follow what's happening in the `-debug=net` log.

Reply:
But see also this discussion on logging: https://github.com/bitcoin/bitcoin/pull/25717#discussion_r936216176

Reply:
The new log stuff is good enough.